cpp/src/io/csv/csv_reader.cu

```diff
@@ -921,60 +855,114 @@ gdf_error getUncompressedHostData(const char* h_data, size_t num_bytes, const st
  * 
  * @param[in] h_uncomp_data Pointer to the uncompressed csv data in host memory
  * @param[in] h_uncomp_size Size of the input data, in bytes
- * @param[in,out] csvData Structure containing the csv parsing parameters
+ * @param[in,out] raw_csv Structure containing the csv parsing parameters
  * and intermediate results
  * 
  * @return gdf_error with error code on failure, otherwise GDF_SUCCESS
  *---------------------------------------------------------------------------**/
-gdf_error uploadDataToDevice(const char* h_uncomp_data, size_t h_uncomp_size, raw_csv_t * raw_csv) {
-	vector<cu_recstart_t> h_rec_starts(raw_csv->num_records + 1);
-	CUDA_TRY( cudaMemcpy(h_rec_starts.data(), raw_csv->recStart, sizeof(cu_recstart_t) * h_rec_starts.size(), cudaMemcpyDefault));
-	
-	// Exclude the rows user chose to skip at the start of the file
-	const gdf_size_type first_row = raw_csv->skiprows + max(raw_csv->header_row, 0l);
-	if (raw_csv->num_records > first_row) {
-		raw_csv->num_records = raw_csv->num_records - (long)first_row;
-	}
-	else {
-		checkError(GDF_FILE_ERROR, "Number of records is too small for the specified skiprows and header parameters");
-	}
+gdf_error uploadDataToDevice(const char *h_uncomp_data, size_t h_uncomp_size,
+                             raw_csv_t *raw_csv) {
+
+  // Exclude the rows that are to be skipped from the start
+  GDF_REQUIRE(raw_csv->num_records > raw_csv->skiprows, GDF_INVALID_API_CALL);
+  const auto first_row = raw_csv->skiprows;
+  raw_csv->num_records = raw_csv->num_records - first_row;
+
+  std::vector<cu_recstart_t> h_rec_starts(raw_csv->num_records);
+  CUDA_TRY(cudaMemcpy(h_rec_starts.data(), raw_csv->recStart + first_row,
+                      sizeof(cu_recstart_t) * h_rec_starts.size(),
+                      cudaMemcpyDefault));
+
+  // Trim lines that are outside range, but keep one greater for the end offset
+  if (raw_csv->byte_range_size != 0) {
+    auto it = h_rec_starts.end() - 1;
+    while (it >= h_rec_starts.begin() &&
+           *it > cu_recstart_t(raw_csv->byte_range_size)) {
+      --it;
+    }
+    if ((it + 2) < h_rec_starts.end()) {
+      h_rec_starts.erase(it + 2, h_rec_starts.end());
+    }
+  }
 
-	// Restrict the rows to nrows if nrows is smaller than the remaining number of rows
-	if (raw_csv->nrows >= 0 && (gdf_size_type)raw_csv->nrows < raw_csv->num_records) {
-		raw_csv->num_records = (gdf_size_type)raw_csv->nrows;
-	}
+  // Discard any blank/empty and comment lines
+  if (raw_csv->opts.skipblanklines ||
+      raw_csv->opts.comment != raw_csv->opts.terminator) {
+    const auto match1 = raw_csv->opts.skipblanklines ? raw_csv->opts.terminator
```

Comment:
To me it looks like these lines have the same effect as 
const auto empty_line_match = raw_csv->opts.skipblanklines ? raw_csv->opts.terminator : '\0';
...
return (h_uncomp_data[i] == empty_line_match ||
                                 h_uncomp_data[i] == raw_csv->opts.comment);

which is, at least to me, much easier to follow.

Reply:
I think that will break unless we add additional check for end-of-file, etc. We don't do any checks against null char in the parser currently. The reader relies on last row/record occurring after the `\n`, which is a null.

Reply:
You already have a check against '\0' if !skipblanklines and comment == '\0'. If this is not intended, some other changes are needed here.

Reply:
Unless I'm missing something, the code tries to ensure that both match1 and match2 are neither `\0`.

Reply:
if skipblanklines is false, and comment char is '\0', match1 will be '\0'.

Reply:
Discussed this offline and realized I missed the condition above:
if (raw_csv->opts.skipblanklines || raw_csv->opts.comment != raw_csv->opts.terminator)
because of this check, the scenario I mentioned would never happen.

Reply:
Added comment for more clarify and also changed latter condition to `raw_csv->opts.comment != '\0'` - a remnant from earlier revision of the code.